### PR TITLE
Add SimMIM

### PR DESCRIPTION
**Pull request by Animatory**

### CODE REVIEW

File: environment.yml [hunk 1]
```diff
@@ -16,7 +16,7 @@ dependencies:
   - cudatoolkit=11.6
   - parameterized=0.8.1
   - pip:
-      - pytorch-lightning==1.8.4.post0
+      - pytorch-lightning==1.9.0
```
Comment (PososikTeam) [inline environment.yml:19]: Did you merge this branch with main?

File: torchok/tasks/simmim.py [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        """Init ClassificationTask.
+
+        Args:
+            hparams: Hyperparameters that set in yaml file.
+            backbone_name: name of the backbone architecture in the BACKBONES registry.
+            pooling_name: name of the backbone architecture in the POOLINGS registry.
+            backbone_params: parameters for backbone constructor.
+            pooling_params: parameters for neck constructor. `in_channels` will be set automatically based on neck or
+                backbone if neck is absent.
+            encoder_stride:
```
Comment (PososikTeam) [inline torchok/tasks/simmim.py:41]: documentation

File: examples/configs/simmim_pretrain_cifar10.yaml [hunk 3]
```diff
@@ -0,0 +1,103 @@
+task:
+  name: SimMIMTask
```
Comment (PososikTeam) [inline examples/configs/simmim_pretrain_cifar10.yaml:2]: Why not train from pretrain weights?

File: torchok/tasks/simmim.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+            nn.Conv2d(
+                in_channels=self.backbone.out_channels,
+                out_channels=self.backbone_stride ** 2 * 3, kernel_size=1),
+            nn.PixelShuffle(self.backbone_stride),
+        )
+
+        self.input_size = self.backbone.img_size[0]
+        self.patch_size = self.backbone.patch_embed.patch_size[0]
+
+        self.rand_size = self.input_size // self.mask_patch_size
```
Comment (PososikTeam) [inline torchok/tasks/simmim.py:70]: Not understandable names, need commets i think or change name

File: torchok/tasks/simmim.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        self.scale = self.mask_patch_size // self.patch_size
```
Comment (PososikTeam) [inline torchok/tasks/simmim.py:71]: scale of what?

File: torchok/tasks/simmim.py [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+        self.hook_handle = self.backbone.patch_embed.register_forward_hook(self.hook)
+        self.backbone.patch_embed.mask_token = nn.Parameter(torch.zeros(1, 1, self.backbone.embed_dim))
+        trunc_normal_(self.backbone.patch_embed.mask_token, mean=0., std=.02)
+
+        mask = torch.zeros(self.rand_size, self.rand_size, dtype=torch.bool, device=self.device, requires_grad=False)
+        mask = mask.repeat_interleave(self.scale, dim=0).repeat_interleave(self.scale, dim=1)
+        self.register_buffer('mask', mask)
+
+    def hook(self, module, inp, out):
```
Comment (PososikTeam) [inline torchok/tasks/simmim.py:84]: documentation

File: torchok/tasks/simmim.py [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+    def _prepare_mask(self):
+        mask = torch.zeros(self.token_count, dtype=torch.bool, device=self.device, requires_grad=False)
+        if self.training:
+            mask_idx = torch.randperm(self.token_count)[:self.mask_count]
+            mask[mask_idx] = True
+
+        mask = mask.reshape((self.rand_size, self.rand_size))
+        mask = mask.repeat_interleave(self.scale, dim=0).repeat_interleave(self.scale, dim=1)
+        self.register_buffer('mask', mask)
```
Comment (PososikTeam) [inline torchok/tasks/simmim.py:108]: You registr `mask` in the constructor, why you need this here?

File: torchok/tasks/simmim.py [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+        return mask
+
+    def forward_with_gt(self, batch: Dict[str, Union[Tensor, int]]) -> Dict[str, Tensor]:
+        """Forward with ground truth labels."""
+        input_data = batch.pop('image')
+
+        mask = self._prepare_mask()
+        z = self.backbone(input_data)
```
Comment (PososikTeam) [inline torchok/tasks/simmim.py:116]: Don't understand where the mask is applied to input_data. May you explain please?

**Comment by Animatory** [comment]
Closed because this method does not provide good results
